benches/hashmap_benchmark.rs

```diff
@@ -0,0 +1,228 @@
+// Insert (which doesn't completely avoid updates, but makes them unlikely),
+// remove and search have benchmarks with empty values and with custom structs
+// of 42 64-bit integers.
+// (as a sidenote, the performance really differs; in the case of remove, the
+// remove function itself returns original value - the benchmark doesn't use
+// this value, but performance is significantly worse - about twice on my
+// machine - than the empty value remove; it might be interesting to see if a
+// remove function returning void would perform better, ie. if the returns
+// are optimized - omitted in this case).
+// The counts of inserted/removed/searched elements are chosen at random from
+// constant ranges in an attempt to avoid a single count performing better
+// because of specific HW features of computers the code is benchmarked with.
+
+extern crate concread;
+extern crate criterion;
+extern crate rand;
+
+use concread::hashmap::*;
+use criterion::{criterion_group, criterion_main, BatchSize, Criterion};
+use rand::{thread_rng, Rng};
+use std::mem;
+
+// ranges of counts for different benchmarks:
+const INSERT_COUNT: (usize, usize) = (120, 140);
+const INSERT_COUNT_FOR_REMOVE: (usize, usize) = (340, 360);
+const REMOVE_COUNT: (usize, usize) = (120, 140);
+const INSERT_COUNT_FOR_SEARCH: (usize, usize) = (120, 140);
+const SEARCH_COUNT: (usize, usize) = (120, 140);
+// In the search benches, we randomly search for elements of a range of SEARCH_SIZE.0/SEARCH_SIZE.1
+// times the number of elements contained.
+const SEARCH_SIZE: (usize, usize) = (4, 3);
+
+pub fn insert_empty_value(c: &mut Criterion) {
+    c.bench_function("insert_empty_value", |b| {
+        b.iter_batched(
+            || {
+                let mut rng = thread_rng();
+                let count = rng.gen_range(INSERT_COUNT.0, INSERT_COUNT.1);
+                let mut list = Vec::with_capacity(count);
+                for _ in 0..count {
+                    list.push(rng.gen_range(0, INSERT_COUNT.1 << 8) as u32)
+                }
+                (HashMap::new(), list)
+            },
+            |mut data| insert_vec(&mut data),
+            BatchSize::SmallInput,
+        )
+    });
+}
+
+pub fn insert_struct_value(c: &mut Criterion) {
+    c.bench_function("insert_struct_value", |b| {
+        b.iter_batched(
+            || {
+                let mut rng = thread_rng();
+                let count = rng.gen_range(INSERT_COUNT.0, INSERT_COUNT.1);
+                let mut list = Vec::with_capacity(count);
+                for _ in 0..count {
+                    list.push((
+                        rng.gen_range(0, INSERT_COUNT.1 << 8) as u32,
+                        default_struct(),
+                    ))
+                }
+                (HashMap::<u32, Struct>::new(), list)
+            },
+            |mut data| insert_struct_vec(&mut data),
+            BatchSize::SmallInput,
+        )
+    });
+}
+
+pub fn remove_empty_value(c: &mut Criterion) {
```

Comment:
Instead of "remove empty value" you could just do `data: &[(K, V)]`, and make `V: ()`, as rust will "compile it away". That way you are expressing the problem a bit more clearly, and it means you can have insert_value which is generic over the two problems (adding with a value and with a unit `()` value).

Reply:
remove_empty_value calls the same prepare_remove and remove_vec that remove_struct_value_no_read calls, so generics are used, but separate functions are still needed for the use of Criterion AFAIK.
Don't think this can be improved much, but please let me know if I'm wrong.